### PR TITLE
build warning fixes

### PR DESCRIPTION
**Pull request by rkhoury**
It's a bit sucky, but the default compilation using gcc seems to be pedantic.  These are fixes for all of the warnings that come up plus some extra fixes for access to inaccessible/protected members (there may be more of the latter).

### CODE REVIEW

File: Source/cmGlobalFastbuildGenerator.cxx [hunk 1]
```diff
@@ -609,7 +609,7 @@ class cmGlobalFastbuildGenerator::Detail::Detection
 			return false;
 		}
 
-		cmTarget::TargetType targetType = target.GetType();
+		//cmTarget::TargetType targetType = target.GetType();
```
Comment (inbilla) [inline Source/cmGlobalFastbuildGenerator.cxx]: Delete this line?

File: Source/cmGlobalFastbuildGenerator.cxx [hunk 2]
```diff
@@ -702,8 +702,7 @@ class cmGlobalFastbuildGenerator::Detail::Detection
 		for (std::vector<std::string>::iterator i = compileCmds.begin();
 			i != compileCmds.end(); ++i)
 		{
-			std::string & compileCmd = *i;
-			lg->ExpandRuleVariables(compileCmd, compileObjectVars);
+			lg->ExpandRuleVariables(*i, compileObjectVars);
```
Comment (inbilla) [inline Source/cmGlobalFastbuildGenerator.cxx]: Why? I prefer to break away from iterators asap. Easier to read

File: Source/cmGlobalFastbuildGenerator.cxx [hunk 3]
```diff
@@ -1036,7 +1037,7 @@ class cmGlobalFastbuildGenerator::Detail::Detection
 			typedef std::vector<std::string> StringVector;
 			typedef std::vector<const TType*> OrderedEntrySet;
 			typedef std::map<std::string, const TType*> OutputMap;
-			typedef std::map<const TType*, StringVector> DependencyMap;
+			//typedef std::map<const TType*, StringVector> DependencyMap;
```
Comment (inbilla) [inline Source/cmGlobalFastbuildGenerator.cxx]: Never used? delete the line

File: Source/cmGlobalFastbuildGenerator.cxx [hunk 4]
```diff
@@ -883,7 +884,7 @@ class cmGlobalFastbuildGenerator::Detail::Detection
 	}
 
 	static void DetectTargetLinkDependencies(
-		cmGlobalFastbuildGenerator* gg,
+		cmGlobalFastbuildGenerator * /*gg*/,
```
Comment (inbilla) [inline Source/cmGlobalFastbuildGenerator.cxx]: Again, never used, just delete the parameter

File: Source/cmGlobalFastbuildGenerator.cxx [hunk 5]
```diff
@@ -379,7 +379,7 @@ class cmGlobalFastbuildGenerator::Detail::Detection
 
 	static void DetectOutput(
 		FastbuildTargetNames & targetNamesOut,
-		cmLocalFastbuildGenerator *lg,
+		cmLocalFastbuildGenerator * /*lg*/,
```
Comment (inbilla) [inline Source/cmGlobalFastbuildGenerator.cxx]: Just delete the parameter

File: Source/cmGlobalFastbuildGenerator.cxx [hunk 6]
```diff
@@ -1903,9 +1915,11 @@ class cmGlobalFastbuildGenerator::Detail::Generation
 		const std::string& targetName = target.GetName();
 
 		// Iterating over all configurations
-		const char* customCommandGroupNamePrefix = "ObjectGroup_cmCustomCommands_";
-		for (std::vector<std::string>::iterator iter = context.self->Configurations.begin();
-			iter != context.self->Configurations.end(); ++iter)
+		//const char* customCommandGroupNamePrefix = "ObjectGroup_cmCustomCommands_";
```
Comment (inbilla) [inline Source/cmGlobalFastbuildGenerator.cxx]: Delete the line?

File: Source/cmGlobalFastbuildGenerator.cxx [hunk 7]
```diff
@@ -2035,10 +2049,12 @@ class cmGlobalFastbuildGenerator::Detail::Generation
...
 		// This time to define linker settings for each config
-		for (std::vector<std::string>::iterator iter = context.self->Configurations.begin();
-			iter != context.self->Configurations.end(); ++iter)
+		std::vector<std::string>::const_iterator
+				configIter = context.self->GetConfigurations().begin(),
+				configEnd = context.self->GetConfigurations().end();
+		for (; configIter != configEnd; ++configIter)
 		{
-			std::string configName = *iter;
+			std::string configName = *configIter;
```
Comment (inbilla) [inline Source/cmGlobalFastbuildGenerator.cxx]: should be const std::string & ?

File: Source/cmGlobalFastbuildGenerator.cxx [hunk 8]
```diff
@@ -2129,10 +2145,10 @@ class cmGlobalFastbuildGenerator::Detail::Generation
 			context.fc.WritePushScopeStruct();
 
 			// Iterating over all configurations
-			for (std::vector<std::string>::iterator iter = context.self->Configurations.begin();
-				iter != context.self->Configurations.end(); ++iter)
+			configIter = context.self->GetConfigurations().begin();
+			for (; configIter != configEnd; ++configIter)
 			{
-				std::string configName = *iter;
+				std::string configName = *configIter;
```
Comment (inbilla) [inline Source/cmGlobalFastbuildGenerator.cxx]: another const std::string&

File: Source/cmGlobalFastbuildGenerator.cxx [hunk 9]
```diff
@@ -2266,8 +2282,10 @@ class cmGlobalFastbuildGenerator::Detail::Generation
...
 		// Iterate over each configuration
 		// This time to define linker settings for each config
-		for (std::vector<std::string>::iterator iter = context.self->Configurations.begin();
-			iter != context.self->Configurations.end(); ++iter)
+		std::vector<std::string>::const_iterator
+				iter = context.self->GetConfigurations().begin(),
+				end = context.self->GetConfigurations().end();
+		for (; iter != end; ++iter)
 		{
 			std::string configName = *iter;
```
Comment (inbilla) [inline Source/cmGlobalFastbuildGenerator.cxx]: And another const std::string & 
:)

File: Source/cmGlobalFastbuildGenerator.cxx [hunk 10]
```diff
@@ -2522,7 +2540,7 @@ class cmGlobalFastbuildGenerator::Detail::Generation
 			}
 
 			cmTarget* target = findResult->second.target;
-			cmLocalFastbuildGenerator* lg = findResult->second.lg;
+			//cmLocalFastbuildGenerator* lg = findResult->second.lg;
```
Comment (inbilla) [inline Source/cmGlobalFastbuildGenerator.cxx]: Delete line?

File: Source/cmGlobalFastbuildGenerator.h [hunk 11]
```diff
@@ -55,6 +55,12 @@ class cmGlobalFastbuildGenerator
 	virtual void ComputeTargetObjectDirectory(cmGeneratorTarget*) const;
 	virtual const char* GetCMakeCFGIntDir() const;
 
+	virtual void GetTargetSets(TargetDependSet& projectTargets,
```
Comment (inbilla) [inline Source/cmGlobalFastbuildGenerator.h:58]: Where did this one come from?

File: Source/cmGlobalFastbuildGenerator.h [hunk 12]
```diff
@@ -55,6 +55,12 @@ class cmGlobalFastbuildGenerator
...
+							   TargetDependSet& originalTargets,
+							   cmLocalGenerator* root, GeneratorVector const&);
+
+	const std::vector<std::string> & GetConfigurations() const;
```
Comment (inbilla) [inline Source/cmGlobalFastbuildGenerator.h:62]: Is this needed?
Everything in the C++ should have 100% access to the members of the global generator...

**Comment by inbilla** [comment]
Just a few comments...
Where did the new functions come from? 
I expected all the variables to be accessible inside the C++...

File: Source/cmGlobalFastbuildGenerator.cxx (hunk 1 above)
Comment (rkhoury) [inline Source/cmGlobalFastbuildGenerator.cxx]: Sure

File: Source/cmGlobalFastbuildGenerator.cxx (hunk 2 above)
Comment (rkhoury) [inline Source/cmGlobalFastbuildGenerator.cxx]: The only reason was the re-use/shadowing of the variable name compileCmd that is defined in an outer scope.  I'm happy to keep this in, I'll just change the name so gcc stops whinging.

File: Source/cmGlobalFastbuildGenerator.cxx (hunk 3 above)
Comment (rkhoury) [inline Source/cmGlobalFastbuildGenerator.cxx]: sure

File: Source/cmGlobalFastbuildGenerator.cxx (hunk 4 above)
Comment (rkhoury) [inline Source/cmGlobalFastbuildGenerator.cxx]: done

File: Source/cmGlobalFastbuildGenerator.cxx (hunk 5 above)
Comment (rkhoury) [inline Source/cmGlobalFastbuildGenerator.cxx]: ok

File: Source/cmGlobalFastbuildGenerator.cxx (hunk 6 above)
Comment (rkhoury) [inline Source/cmGlobalFastbuildGenerator.cxx]: why not!

File: Source/cmGlobalFastbuildGenerator.cxx (hunk 7 above)
Comment (rkhoury) [inline Source/cmGlobalFastbuildGenerator.cxx]: As you wish. I was just keeping with the previous code (I really was only concerned about the warnings going away)

File: Source/cmGlobalFastbuildGenerator.cxx (hunk 8 above)
Comment (rkhoury) [inline Source/cmGlobalFastbuildGenerator.cxx]: done

File: Source/cmGlobalFastbuildGenerator.cxx (hunk 9 above)
Comment (rkhoury) [inline Source/cmGlobalFastbuildGenerator.cxx]: sure thing

File: Source/cmGlobalFastbuildGenerator.cxx (hunk 10 above)
Comment (rkhoury) [inline Source/cmGlobalFastbuildGenerator.cxx]: okie dokie

File: Source/cmGlobalFastbuildGenerator.h (hunk 12 above)
Comment (rkhoury) [inline Source/cmGlobalFastbuildGenerator.h:62]: Not technically, at least not in C++03.
From http://www.open-std.org/jtc1/sc22/wg21/docs/cwg_defects.html#45

> ...
> and 11.7 [class.access.nest] paragraph 1:
> The members of a nested class have no special access to members of an enclosing class,
> nor to classes or functions that have granted friendship to an enclosing class; the usual
> access rules (clause 11 [class.access] ) shall be obeyed. [...]
> ...

It seems this was changed in c++11, but clang warns when compiling with c++03 support.

**Comment by rkhoury** [comment]
Sorry for closing it off.  I didn't realise that the pull request auto-updated with a new push to the branch.

**Comment by inbilla** [comment]
All good, I think "closing" means - "I don't want you to pull yet"
seems to keep history

**Comment by rkhoury** [comment]
I assume it compiled all fine in VS?

**Pull request merged**